src/Domain.LinnApps/Reports/StoragePlaceAuditReportService.cs

```diff
@@ -0,0 +1,151 @@
+﻿namespace Linn.Stores2.Domain.LinnApps.Reports
+{
+    using System.Collections.Generic;
+    using System.Linq;
+
+    using Linn.Common.Persistence;
+    using Linn.Common.Reporting.Models;
+    using Linn.Stores2.Domain.LinnApps.Stock;
+
+    public class StoragePlaceAuditReportService : IStoragePlaceAuditReportService
+    {
+        private readonly IReportingHelper reportingHelper;
+
+        private readonly IRepository<StockLocator, int> stockLocatorRepository;
+
+        public StoragePlaceAuditReportService(
+            IReportingHelper reportingHelper,
+            IRepository<StockLocator, int> stockLocatorRepository)
+        {
+            this.reportingHelper = reportingHelper;
+            this.stockLocatorRepository = stockLocatorRepository;
+        }
+
+        public ResultsModel StoragePlaceAuditReport(IEnumerable<string> locationList, string locationRange)
+        {
+            var stockLocators = new List<StockLocator>();
+
+            if (!string.IsNullOrEmpty(locationRange) && locationRange.StartsWith("E-K"))
+            {
+                stockLocators = this.stockLocatorRepository
+                    .FilterBy(s => s.CurrentStock == "Y" && s.StorageLocation.LocationCode.StartsWith(locationRange))
+                    .OrderBy(s => s.StorageLocation.LocationCode).ThenBy(s => s.PartNumber).ToList();
+            }
+            else
+            {
+                foreach (var loc in locationList)
+                {
+                    if (loc.StartsWith("P"))
+                    {
+                        stockLocators.AddRange(this.stockLocatorRepository
+                            .FilterBy(s => s.CurrentStock == "Y" && int.Parse(loc.Substring(1)) == s.PalletNumber));
+                    }
+                    else
+                    {
+                        stockLocators.AddRange(this.stockLocatorRepository
+                            .FilterBy(s => s.CurrentStock == "Y" && loc == s.StorageLocation.LocationCode));
+                    }
+                }
+
+                stockLocators = stockLocators
+                    .OrderBy(a => a.PalletNumber)
+                    .ThenBy(b => b.StorageLocation?.LocationCode)
+                    .ThenBy(c => c.PartNumber).ToList();
```

Comment:
could ToListAsync again -  Almost definiitely won't noticve any benefit though other than freeing up a thread while the query is executed, but just in case you wanted to give it a try

Reply:
actually on second thoughts I think this would require a bit of adjusting to your tests too so feel free to ignore me. I had intended to have a quick chat about all this stuff at our next show & tell anyways